model/project_parser_storage.go

```diff
@@ -77,3 +78,40 @@ func ParserProjectUpsertOne(ctx context.Context, settings *evergreen.Settings, m
 	defer ppStorage.Close(ctx)
 	return ppStorage.UpsertOne(ctx, pp)
 }
+
+// ParserProjectUpsertOneWithS3Fallback attempts to upsert the parser project
+// into persistent storage using the given storage method. If it fails due to
+// DB document size limitations, it will attempt to fall back to using S3
+// to store it. If it succeeds, this returns the actual project storage method
+// used to persist the parser project; otherwise, it returns the
+// originally-requested storage method.
+func ParserProjectUpsertOneWithS3Fallback(ctx context.Context, settings *evergreen.Settings, method evergreen.ParserProjectStorageMethod, pp *ParserProject) (evergreen.ParserProjectStorageMethod, error) {
+	err := ParserProjectUpsertOne(ctx, settings, method, pp)
+	if err == nil {
+		return method, nil
+	}
+	if method == evergreen.ProjectStorageMethodS3 {
+		return method, err
+	}
+
+	if !db.IsDocumentLimit(err) {
+		return method, err
+	}
+
+	// Upsertion could fail for multiple reasons, but Evergreen only resorts to
+	// using S3 if upsert failed due to the DB document size limitation.
+
```

Comment:
I intentionally made it a floating comment because it's meant to apply to everything below this line. I'm just gonna remove it since the doc comment already covers when it would fall back to S3.